File: 백준/레벨Gold V/1717. 집합의 표현/집합의 표현.py
from sys import stdin

input = stdin.readline

def disjoint_set():
    # 부모 노드 초기화
    N, M = map(int, input().split())
    parent = list(range(N + 1))

    def find(x):
        # 경로 압축
        while parent[x] != x:
            parent[x] = parent[parent[x]]
            x = parent[x]
        return x
    
    def union(a, b):
        root_a = find(a)
        root_b = find(b)
        
        if root_a > root_b:
            parent[root_a] = root_b
        else:
            parent[root_b] = root_a

    # 서로소 집합 연산
    for _ in range(M):
        op, a, b = map(int, input().split())
        
        if op == 0:
            union(a, b)
        elif find(a) == find(b):
            print("YES")
        else:
            print("NO")

disjoint_set()

File: 백준/레벨Gold V/1717. 집합의 표현/집합의 표현.ts
// @ts-ignore
import fs = require('fs');

const read = fs.readFileSync(0, 'utf-8');
const lines = read.split('\n').map((line: string) => line.split(' ').map(Number));

const disjointSet = () => {
  // 부모 노드 초기화
  const [N, M] = lines[0];
  const parent = Array.from({ length: N + 1 }, (_, i) => i);

  const find = (x: number): number => {
    // 경로 압축
    while (parent[x] !== x) {
      parent[x] = parent[parent[x]];
      x = parent[x];
    }
    return x;
  };

  const union = (a: number, b: number): void => {
    const rootA = find(a);
    const rootB = find(b);

    if (rootA > rootB) {
      parent[rootA] = rootB;
    } else {
      parent[rootB] = rootA;
    }
  };

  // 서로소 집합 연산
  const result: string[] = [];
  for (let i = 1; i <= M; i++) {
    const [op, a, b] = lines[i];

    if (op === 0) {
      union(a, b);
    } else {
      result.push(find(a) === find(b) ? 'YES' : 'NO');
    }
  }

  console.log(result.join('\n'));
};

disjointSet();